Source/drlg_l3.cpp

```diff
@@ -1433,74 +1377,113 @@ bool Spawn(int x, int y, int *totarea)
 	return false;
 }
 
-bool HivePlaceSetRandom(const BYTE *miniset, int rndper)
+bool PlaceMiniSet(const Miniset &miniset, bool setview)
 {
-	bool placed = false;
-	int sw = miniset[0];
-	int sh = miniset[1];
+	int sw = miniset.size.width;
+	int sh = miniset.size.height;
+	int sx = GenerateRnd(DMAXX - sw) - 1;
+	int sy = GenerateRnd(DMAXY - sh);
+
+	for (int bailcnt = 0;; bailcnt++) {
+		if (bailcnt > 198)
+			return false;
+
+		sx++;
+		if (sx == DMAXX - sw) {
+			sx = 0;
+			sy++;
+			if (sy == DMAXY - sh) {
+				sy = 0;
+			}
+		}
+
+		if (SetPiecesRoom.Contains({ sx, sy }))
+			continue;
+		if (miniset.matches({ sx, sy }))
+			break;
+	}
 
+	miniset.place({ sx, sy });
+
+	if (setview) {
+		ViewPosition = Point { 17, 19 } + Displacement { sx, sy } * 2;
+	}
+
+	return true;
+}
+
+bool CanReplaceTile(uint8_t replace, Point tile)
+{
+	if (replace < 84 || replace > 100) {
+		return true;
+	}
+
+	// BUGFIX: p2 is a workaround for a bug, only p1 should have been used (fixing this breaks compatability)
+	constexpr auto ComparisonWithBoundsCheck = [](Point p1, Point p2) {
+		return (p1.x >= 0 && p1.x < DMAXX && p1.y >= 0 && p1.y < DMAXY)
+		    && (p2.x >= 0 && p2.x < DMAXX && p2.y >= 0 && p2.y < DMAXY)
+		    && (dungeon[p1.x][p1.y] >= 84 && dungeon[p2.x][p2.y] <= 100);
+	};
+	if (ComparisonWithBoundsCheck(tile + Direction::NorthWest, tile + Direction::NorthWest)
+	    || ComparisonWithBoundsCheck(tile + Direction::SouthEast, tile + Direction::NorthWest)
+	    || ComparisonWithBoundsCheck(tile + Direction::SouthWest, tile + Direction::NorthWest)
+	    || ComparisonWithBoundsCheck(tile + Direction::NorthEast, tile + Direction::NorthWest)) {
+		return false;
+	}
+
+	return true;
+}
+
+bool PlaceMiniSetRandom(const Miniset &miniset, int rndper)

```

Comment:
A comment to describe the behaviour of this function would be good. Something like `/* Randomly places the given miniset throughout the dungeon wherever it would fit, returning true if at least one instance was placed */`?